File: src/main/java/no/nav/syfo/consumer/ws/ArbeidsfordelingConsumer.java
package no.nav.syfo.consumer.ws;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.domain.dto.Soknadstype;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.ArbeidsfordelingV1;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.FinnBehandlendeEnhetListeUgyldigInput;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.*;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.WSFinnBehandlendeEnhetListeRequest;
import org.springframework.stereotype.Component;

import static no.nav.syfo.domain.dto.Soknadstype.OPPHOLD_UTLAND;
import static no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.WSEnhetsstatus.AKTIV;

@Component
@Slf4j
public class ArbeidsfordelingConsumer {

    private ArbeidsfordelingV1 arbeidsfordelingV1;

    public static final String BEHANDLINGSTEMA_OPPHOLD_UTLAND = "ab0314";

    public ArbeidsfordelingConsumer(ArbeidsfordelingV1 arbeidsfordelingV1) {
        this.arbeidsfordelingV1 = arbeidsfordelingV1;
    }

    public String finnBehandlendeEnhet(GeografiskTilknytning geografiskTilknytning, Soknadstype soknadstype) {
        try {
            return arbeidsfordelingV1.finnBehandlendeEnhetListe(new WSFinnBehandlendeEnhetListeRequest()
                    .withArbeidsfordelingKriterier(new WSArbeidsfordelingKriterier()
                            .withDiskresjonskode(geografiskTilknytning.diskresjonskode != null ? new WSDiskresjonskoder().withValue(geografiskTilknytning.diskresjonskode) : null)
                            .withGeografiskTilknytning(new WSGeografi().withValue(geografiskTilknytning.geografiskTilknytning))
                            .withTema(new WSTema().withValue("SYK"))
                            .withBehandlingstema(hentRiktigTemaBehandlingstemaForSoknadstype(soknadstype))))
                    .getBehandlendeEnhetListe()
                    .stream()
                    .filter(wsOrganisasjonsenhet -> AKTIV.equals(wsOrganisasjonsenhet.getStatus()))
                    .map(WSOrganisasjonsenhet::getEnhetId)
                    .findFirst()
                    .orElseThrow(() -> new RuntimeException("Fant ingen aktiv enhet for " + geografiskTilknytning));
        } catch (FinnBehandlendeEnhetListeUgyldigInput e) {
            log.error("Feil ved henting av brukers forvaltningsenhet", e);
            throw new RuntimeException("Feil ved henting av brukers forvaltningsenhet", e);
        } catch (RuntimeException e) {
            log.error("Klarte ikke å hente behandlende enhet!", e);
            throw new RuntimeException(e);
        }
    }

    private WSBehandlingstema hentRiktigTemaBehandlingstemaForSoknadstype(Soknadstype soknadstype) {
        if (soknadstype == OPPHOLD_UTLAND) {
            return new WSBehandlingstema().withValue(BEHANDLINGSTEMA_OPPHOLD_UTLAND);
        }
        return null;
    }
}

File: src/main/kotlin/no/nav/syfo/consumer/ws/ArbeidsfordelingConsumer.kt
package no.nav.syfo.consumer.ws

import no.nav.syfo.domain.dto.Soknadstype
import no.nav.syfo.domain.dto.Soknadstype.OPPHOLD_UTLAND
import no.nav.syfo.logger
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.binding.ArbeidsfordelingV1
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.binding.FinnBehandlendeEnhetListeUgyldigInput
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.ArbeidsfordelingKriterier
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Behandlingstema
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Diskresjonskoder
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Enhetsstatus
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Geografi
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Tema
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.FinnBehandlendeEnhetListeRequest
import org.springframework.stereotype.Component

const val BEHANDLINGSTEMA_OPPHOLD_UTLAND = "ab0314"
const val NAV_OPPFOLGING_UTLAND_KONTOR_NR = "0393"

@Component
class ArbeidsfordelingConsumer(private val arbeidsfordelingV1: ArbeidsfordelingV1) {
    private val log = logger()

    fun finnBehandlendeEnhet(tilknytning: GeografiskTilknytning, soknadstype: Soknadstype): String {
        try {
            return if ("NOR" == tilknytning.geografiskTilknytning)
                "4474"
            else
                arbeidsfordelingV1.finnBehandlendeEnhetListe(
                    FinnBehandlendeEnhetListeRequest().apply {
                        arbeidsfordelingKriterier = ArbeidsfordelingKriterier().apply {
                            diskresjonskode = if (tilknytning.diskresjonskode == null) null else Diskresjonskoder().apply {
                                value = tilknytning.diskresjonskode
                            }
                            geografiskTilknytning = Geografi().apply {
                                value = tilknytning.geografiskTilknytning
                            }
                            tema = Tema().apply {
                                value = "SYK"
                            }
                            behandlingstema = hentRiktigTemaBehandlingstemaForSoknadstype(soknadstype)
                        }
                    }
                )
                    .behandlendeEnhetListe
                    .firstOrNull { organisasjonsenhet -> organisasjonsenhet.status == Enhetsstatus.AKTIV }?.enhetId
                    ?: throw RuntimeException("Fant ingen aktiv enhet for ${tilknytning.geografiskTilknytning}")
        } catch (e: FinnBehandlendeEnhetListeUgyldigInput) {
            log.error("Feil ved henting av brukers forvaltningsenhet", e)
            throw RuntimeException("Feil ved henting av brukers forvaltningsenhet", e)
        } catch (e: RuntimeException) {
            log.warn(
                "Klarte ikke å hente behandlende enhet! Gir oppgaven til NAV_OPPFOLGING_UTLAND ($NAV_OPPFOLGING_UTLAND_KONTOR_NR)",
                e
            )
            return NAV_OPPFOLGING_UTLAND_KONTOR_NR
        }
    }

    private fun hentRiktigTemaBehandlingstemaForSoknadstype(soknadstype: Soknadstype?): Behandlingstema? {
        return if (soknadstype === OPPHOLD_UTLAND) {
            Behandlingstema().apply {
                value = BEHANDLINGSTEMA_OPPHOLD_UTLAND
            }
        } else null
    }
}
